src/lib.rs

```diff
@@ -71,40 +80,52 @@ impl MsgEnvelope {
         self.message.id()
     }
 
-    /// This is not quite good.
-    /// It does work for the cases we have,
-    /// but it does so without being clearly robust/flexible.
-    /// So, needs some improvement..
+    /// Verify the signature provided by the most recent sender is valid.
     pub fn verify(&self) -> Result<bool> {
         let data = if self.proxies.is_empty() {
-            bincode::serialize(&self.message)
-                .map_err(|_| Error::Serialisation("Could not serialize message".to_string()))?
+            self.serialise()?
         } else {
             let mut msg = self.clone();
             let _ = msg.proxies.pop();
-            bincode::serialize(&msg)
-                .map_err(|_| Error::Serialisation("Could not serialize message".to_string()))?
+            msg.serialise()?
         };
+
         let sender = self.most_recent_sender();
         Ok(sender.verify(&data))
     }
 
+    /// Deserialise a MsgEnvelope from bytes received over the wire.
+    pub fn from(bytes: &[u8]) -> Result<Self> {
```

Comment:
should we not use `bytes::Bytes` here? (if not, why not?)

Reply:
yeah, I think we should